### ISSUE TITLE
[2.4.0.RC2-1] CoreDslMacros.errorImpl doesn't accept partial function with unapply method

### ISSUE THREAD

**Issue by seratch** [scalatra/scalatra#527]
https://github.com/scalatra/scalatra/pull/526/files#diff-7f6d1b50cff6ec9e973a74b31052bb55R43

```scala
error {
  case scala.util.control.NonFatal(e) =>
}
```

This code fails to compile as below:

```
[error] unexpected UnApply scala.util.control.NonFatal.unapply(<unapply-selector>) <unapply> ((e @ _))
scala.reflect.internal.FatalError: unexpected UnApply scala.util.control.NonFatal.unapply(<unapply-selector>) <unapply> ((e @ _))
	at scala.reflect.internal.Reporting$class.abort(Reporting.scala:59)
	at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:16)
	at scala.tools.nsc.typechecker.Typers$Typer.typedInAnyMode$1(Typers.scala:5348)
	at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5360)
	at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5396)
	at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5423)
	at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5370)
	at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5374)
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$118$$anonfun$apply$58.apply(Typers.scala:5509)
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$118.apply(Typers.scala:5509)
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$118.apply(Typers.scala:5509)
	at scala.tools.nsc.typechecker.TypeDiagnostics$class.typingInPattern(TypeDiagnostics.scala:62)
	at scala.tools.nsc.Global$$anon$1.typingInPattern(Global.scala:463)
	at scala.tools.nsc.typechecker.Typers$Typer.typedPattern(Typers.scala:5509)
	at scala.tools.nsc.typechecker.Typers$Typer.typedCase(Typers.scala:2400)
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$typedCases$1.apply(Typers.scala:2440)
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$typedCases$1.apply(Typers.scala:2439)
	at scala.collection.immutable.List.loop$1(List.scala:173)
	at scala.collection.immutable.List.mapConserve(List.scala:189)
	at scala.tools.nsc.typechecker.Typers$Typer.typedCases(Typers.scala:2439)
	at scala.tools.nsc.typechecker.Typers$Typer.typedMatch(Typers.scala:2451)
	at scala.tools.nsc.typechecker.Typers$Typer.typedVirtualizedMatch$1(Typers.scala:4309)
	at scala.tools.nsc.typechecker.Typers$Typer.typedOutsidePatternMode$1(Typers.scala:5322)
	at scala.tools.nsc.typechecker.Typers$Typer.typedInAnyMode$1(Typers.scala:5353)
	at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5360)
	at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5396)
	at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5423)
	at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5370)
	at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5374)
	at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5463)
	at scala.tools.nsc.typechecker.Typers$Typer.computeType(Typers.scala:5554)
	at scala.tools.nsc.typechecker.Namers$Namer.assignTypeToTree(Namers.scala:879)
	at scala.tools.nsc.typechecker.Namers$Namer.methodSig(Namers.scala:1213)
	at scala.tools.nsc.typechecker.Namers$Namer.getSig$1(Namers.scala:1529)
	at scala.tools.nsc.typechecker.Namers$Namer.typeSig(Namers.scala:1541)
	at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply$mcV$sp(Namers.scala:781)
	at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply(Namers.scala:780)
	at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply(Namers.scala:780)
	at scala.tools.nsc.typechecker.Namers$Namer.scala$tools$nsc$typechecker$Namers$Namer$$logAndValidate(Namers.scala:1568)
	at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1.apply(Namers.scala:780)
	at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1.apply(Namers.scala:772)
	at scala.tools.nsc.typechecker.Namers$$anon$1.completeImpl(Namers.scala:1684)
	at scala.tools.nsc.typechecker.Namers$LockingTypeCompleter$class.complete(Namers.scala:1692)
	at scala.tools.nsc.typechecker.Namers$$anon$1.complete(Namers.scala:1682)
	at scala.tools.nsc.typechecker.Namers$PolyTypeCompleter.completeImpl(Namers.scala:1725)
	at scala.tools.nsc.typechecker.Namers$LockingTypeCompleter$class.complete(Namers.scala:1692)

```

**Comment by dozed** [comment]
This seems to be a general problem with `untypecheck`. I asked on the mailing list: https://groups.google.com/forum/#!topic/scala-language/ihQqfn_Fpp8

**Comment by seratch** [comment]
Fixed by #524
